### PR TITLE
Support multiple insertion text property names

### PR DESCRIPTION
**Pull request by daytonellwanger**
This change fixes a bug where if the guest invokes completion in a Live Share session in a ts file and selects an IntelliCode completion, the star will also be inserted (e.g.  console.★ log).

We were previously trying to extract the insertion text on the completion item by looking at the "InsertionText" property. TS IntelliCode uses "InsertionText" for their property name.

This same change was made in the Live Share code base and was never migrated to Roslyn. See [here](https://devdiv.visualstudio.com/DevDiv/_git/Cascade/pullrequest/189799?_a=overview).

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+
+    internal static class CompletionItemExtensions
+    {
+
+        // A list of the different property names that the insertion text of a completion item may be stored under.
```
Comment (daytonellwanger) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: @heejaechang @dibarbet I'm sure there's a better way to implement this functionality now that this code lives in the Roslyn code base. Let me know!

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+
+    internal static class CompletionItemExtensions
+    {
+
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: spacing nit
```suggestion
```

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        internal static string GetInsertionText(this CompletionItem item)
+        {
+            foreach (var insertionTextPropertyName in InsertionTextPropertyNames)
+            {
+                if (item.Properties.ContainsKey(insertionTextPropertyName))
+                {
+                    return item.Properties[insertionTextPropertyName];
+                }
+            }
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: spacing nit
```suggestion
            }
            
```

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        // A list of the different property names that the insertion text of a completion item may be stored under.
+        private static readonly string[] InsertionTextPropertyNames = new string[] { "InsertionText", "InsertText" };
+
+        internal static string GetInsertionText(this CompletionItem item)
+        {
+            foreach (var insertionTextPropertyName in InsertionTextPropertyNames)
+            {
+                if (item.Properties.ContainsKey(insertionTextPropertyName))
+                {
+                    return item.Properties[insertionTextPropertyName];
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: I'm not super familiar with roslyn completetion, tagging @heejaechang
Is the display text ever different from the insertion text?  It seems like we just set the insertion text to the display text - http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs,368
but I might be wrong.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Completion;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]:
```suggestion
```

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    internal static class CompletionItemExtensions
+    {
+
+        // A list of the different property names that the insertion text of a completion item may be stored under.
+        private static readonly string[] InsertionTextPropertyNames = new string[] { "InsertionText", "InsertText" };
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: this seems super strange.  how woudl this have worked before?

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: i.e. i don't believe roslyn ever supported both of these strings.  so i don't see how it makes sense to add support for this other string, instead of making it so that whoever creates that string just updates to using the right string instead.

Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Switch to ImmutableArray.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+        internal static string GetInsertionText(this CompletionItem item)
+        {
+            foreach (var insertionTextPropertyName in InsertionTextPropertyNames)
+            {
+                if (item.Properties.ContainsKey(insertionTextPropertyName))
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: can you use TryGetValue instead so you don' tneed multiple lookups.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: We often set the insertion text differently than display text.   We only use the display text if there is no insertion text.

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+                {
+                    return item.Properties[insertionTextPropertyName];
+                }
+            }
+            return item.DisplayText;
+        }
+    }
+
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]:
```suggestion
```

**Comment by CyrusNajmabadi** [comment]
> TS IntelliCode uses "InsertionText" for their property name.

Why don't we fix things up to use the same name.  It's unclear to me why a different name was ever used, or how that would have been expected to work in the first palce as Roslyn woudl have only looked for 'InsertionText' and not 'InsertText'.

**Comment by CyrusNajmabadi** [comment]
> This same change was made in the Live Share code base and was never migrated to Roslyn. See here.

I can't see the linked issue.  Can you summarize/copy it here?  Thanks!

File: src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs (hunk 6 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: generally in agreement with @CyrusNajmabadi here, why is it in some cases (c#) intellicode leave insertion text as-is, but for TS they include the star?

If I had a guess it would be that for TS they don't specify the insertion text (and instead specify the 'InsertText' so it falls back to the display text.
If we can have TS specify the 'InsertionText' property, that would be my preferable approach.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: They do specify an insertion text, but the property name they use is "InsertText," not "InsertionText." If this isn't something ya'll are aware of Roslyn explicitly supporting, I'm in agreement with both of you - this should be fixed by whoever is using this non-standard insert text property, and not by Roslyn.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: > this should be fixed by whoever is using this non-standard insert text property, and not by Roslyn.

Agreed.  I don't see how `"InsertText"` could have ever worked.  So adding support for this in roslyn to make something work that couldn't have worked before doesn't seem to make sense to me.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: I'm going to do some more investigation and work with the IntelliCode and TypeScript teams as needed. I'll follow up.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: image

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: So after invoking `CompletionService.GetCompletionsAsync`, the items contributed by IntelliCode have an `"InsertText"` property but no `"InsertionText"` property. So we're in agreement this is a TS/IntelliCode bug?

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Yes.  I'm not sure what htey were hoping to accomplish with that given that roslyn would never look at that property.  Also, as per work i did about a year ago, the display text should not be `"<star> log"` it should just be `"log"`.  The `<star>` should be in `DisplayTextPrefix`.  Note: i believe @dpoeschl communicated this to them back when that work happened.  So might as well fix this all up at the same time.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Yes, outside of a Live Share session the insertion text is just `log`. But Live Share surfaces the completion items through a different mechanism than Roslyn (though the source of the completion items is Roslyn), that doesn't have the filtering you implemented, so we get the star included.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Note: this will also impact Roslyn when you move to using the VS LSP client to handle presenting completion items.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Sorry, can you explain to me the architectural changes happening?  The picture above shows a Roslyn completion item.  So something is def populating it :)   I'm just pointing out that they're populating it *wrong* :D  

> Note: this will also impact Roslyn when you move to using the VS LSP client to handle presenting completion items.

Can you explain what the stack will look like?  We just did the work to switch to the new async completion presenter system.  Will we be moving off of that into a new LSP based presenter?

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: > 
> Can you explain what the stack will look like? We just did the work to switch to the new async completion presenter system. Will we be moving off of that into a new LSP based presenter?

Long term the goal for everything is LSP yes, but I think @daytonellwanger means that in the near term, we plan to move **LiveShare** from their custom extensibility point to the VS LSP client ( so we can remove all the liveshare client code).

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Basically, is roslyn involved in that stack at all?  if so, we should make sure the data is going into roslyn in the expected fashion (i.e. `InsertionText` not `InsertText` and not having the `<star>` be in the display text).  If roslyn isn't in the stack... well, then i'm not sure what would need to happen here at all :)

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Ahh, yes roslyn would be providing the server side data in either case so yes we should get it populated correctly.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Yes, sorry for the confusion. Roslyn will still be serving the completion items (so yes, we're all in agreement that those items should have the right shape). Today, Roslyn (or maybe it's the async completion presenter system now) is responsible for presenting the completion items and inserting them when one is selected, and this works; i.e. the star isn't inserted. When that switches to the VS LSP client being responsible for inserting the completion items, this will break (star will be inserted). This is why Live Share is currently broken.

So my point was that TS/IntelliCode should fix this (use "InsertionText," not "InsertText") 😄 .

Today it's only impacting users in a Live Share session, but in the future it may also impact users outside of one.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Spoke with @amcasey, and the TypeScript team is going to switch to using "InsertionText". I'll abandon this PR.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: Great.  Thanks! :)

Comment (amcasey) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: We'll try it - it may not work.  Our property value may not have the same meaning as Roslyn's or this may break us in non-Live Share scenarios.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: I'm definitely a little confused here about what's going on and all teh moving parts.  I guess i should ask: 

1. what is the end goal?
2. what is the core user problem we're seein now?
3. how do the individual pieces connect together and lead up to that problem?

Based on this, we can determine the "right fix".  it seems very strange *currently* to propose any change in roslyn due to something that TypeScript is doing for their own purposes.   Either TS is putting in this field with the expectation that Roslyn will respect it.  or TS is using it for some other purpose, in which case the text should *not* change.

INstead of just trying to slam in stuff into the code, can we step back and identify the problem all the way from the root cause to the end place where it impacts the user?

Thanks!

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: 1. TS IntelliCode completion items are correctly inserted in a Live Share session.
2. If the guest in a Live Share session tries to insert a TS IntelliCode completion item, the "IntelliCode star" will also be inserted.
3. VS LSP client issues completion request to Live Share "language server" -> Live Share guest issues a completion request to Live Share host -> Live Share host asks Roslyn Live Share extension for completion items -> Roslyn Live Share extension asks Roslyn for completion items -> Roslyn asks TypeScript for completion items -> TypeScript returns completion items, with "InsertText" property -> Roslyn Live Share extension maps Roslyn completion items to LSP completion items. Uses display text for insert text because "InsertionText" property is absent -> LSP completion items returned to Live Share guest -> Completion items rendered by VS LSP client. "IntelliCode star" is inserted because the insert text is equal to the display text.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: [This (the Roslyn Live Share handler for completion)](https://github.com/dotnet/roslyn/blob/14db256cb9a7bc4131c2c4ad1f2d109e9055854a/src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs#L66) may be helpful to make (3) above more concrete.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: That's very helpful.  thanks.  The crux of the issue seems, to me, to be:

> Roslyn Live Share extension asks Roslyn for completion items -> Roslyn asks TypeScript for completion items

If this is hte case, then there's a contract between Roslyn and TS that needs to be maintained.  In this case, the expectation is:

1. `<stars>` should not go in the `DisplayText ` property. They should go into the `DisplayTextPrefix`
2. By default roslyn will insert the `DisplayText` into the document.  If a language wants teh DisplayText and inserted text to differ they can do that by overriding https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/Completion/CompletionService.cs#L123

No language should be manually setting the `InsertionText` property.  It is an undocumented implementation detail that is not guaranteed to be supported.  For example, roslyn may stop using that at any time. 

In fact, i don't know why the code you linked does this: https://github.com/dotnet/roslyn/blob/14db256cb9a7bc4131c2c4ad1f2d109e9055854a/src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs#L66

It's not correct.  It's poking into internal impl details that should not be depended on.  The right way to determine what will actually be inserted into the document is *only* to call `CompletionService.GetChangeAsync`.  Note: if you do not do this then *many* completion scenarios will be broken and you will be inserting something that is not what roslyn is expecting to be inserted.

Comment (daytonellwanger) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: @amcasey brought up the same concerns 😄 . It seems like today the Roslyn Live Share extension has a busted completion handler (i.e. it depends on internal implementation details and has different behavior from what Roslyn would insert - e.g. it doesn't use `CompletionService.GetChangeAsync`). Part of this is due to trying to map the way Roslyn handles completion onto the way completion is defined in the LSP. Resolving that will be a longer/more complicated discussion. In the short term, it seems like we could fix my original concern by having the `<star>` go in `DisplayTextPrefix` instead of `DisplayText`.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: > Part of this is due to trying to map the way Roslyn handles completion onto the way completion is defined in the LSP. Resolving that will be a longer/more complicated discussion. In the short term, it seems like we could fix my original concern by having the <star> go in DisplayTextPrefix instead of DisplayText.

That works for me :)    If/when more discussions happen about these different completoin systems, feel free to rope me in.  It's def a complex area.  Unfortunately, and i know how i goes with pressure to ship, there's often a push to do things in more a quick and dirty fashion.   However, that way lies enormous technical dept that just ends up costing more in the long run IMO.   :)

Thanks for all the explanations, it was super super helpful.

Comment (amcasey) [reply src/Features/LanguageServer/Protocol/Handler/Completion/CompletionItemExtensions.cs]: FYI @minestarks
